banzai/main.py

```diff
@@ -419,3 +420,66 @@ def update_db():
         dbs.populate_instrument_tables(db_address=args.db_address)
     except Exception:
         logger.error('Could not populate instruments table: {error}'.format(error=logs.format_exception()))
+
+
+RETRY_DELAY = int(os.getenv('RETRY_DELAY', 1000*60*10))
+
+
+@dramatiq.actor(max_retries=0, queue_name=settings.REDIS_QUEUE_NAMES['SCHEDULE_STACK'])
+def should_retry_schedule_stack(message_data, exception_data):
+    if message_data['options']['retries'] >= 2:
+        schedule_stack(*message_data['args'], process_any_images=True)
+
+
+@dramatiq.actor(max_retries=3, min_backoff=RETRY_DELAY, max_backoff=RETRY_DELAY, queue_name=settings.REDIS_QUEUE_NAMES['SCHEDULE_STACK'])
+def schedule_stack(runtime_context_json, blocks, calibration_type, site, camera, enclosure, telescope, process_any_images=True):
+    logger.info('schedule stack for matching blocks')
+    runtime_context_json['min_date'] = datetime.strptime(runtime_context_json['min_date'], '%Y-%m-%d %H:%M:%S')
+    runtime_context_json['max_date'] = datetime.strptime(runtime_context_json['max_date'], '%Y-%m-%d %H:%M:%S')
+    runtime_context = Context(runtime_context_json)
+    instrument = dbs.query_for_instrument(runtime_context.db_address, site, camera, enclosure, telescope)
+    completed_image_count = len(dbs.get_individual_calibration_images(instrument, calibration_type,
+                                                                      runtime_context.min_date,
+                                                                      runtime_context.max_date,
+                                                                      use_masters=False,
+                                                                      db_address=runtime_context.db_address))
+    expected_image_count = 0
+    for block in blocks:
+        for molecule in block['molecules']:
+            if calibration_type.upper() == molecule['type']:
+                expected_image_count += molecule['exposure_count']
+    logger.info('expected image count: {0}'.format(str(expected_image_count)))
+    if (expected_image_count < completed_image_count and not process_any_images):
+        raise
+    else:
+        process_master_maker(runtime_context, instrument, calibration_type, runtime_context.min_date,
```

Comment:
Shouldn't this be `process_master_maker.send(...`

Reply:
process_master_maker is already being called from an asynchronous worker process, so used .send() would spawn off yet another thread. Probably best not to do that.